#need to make "ind_file" for structure. This is an ordred list of sample names in column 1 with state names (populations) in column 2.
make_ind_file_structure.r

#upload the structure formatted vcfs to server in their respective dirs, Nf_SPANDx_all_seqs_structure_th and Nd_SPANDx_all_seqs_structure_th
# The ind_files are in the repo
#Then perform format conversions with pgspider

module purge
module load linuxbrew/colsa
cd Nf_SPANDx_all_seqs_structure_th
cp ~

#First run pgdspider with no -spid to generate template
bcftools view FINAL_snp.mac_ge2.biallele.LD.structure_analyses.vcf.gz > FINAL_snp.mac_ge2.biallele.LD.structure_analyses.vcf

pgdspider -inputfile FINAL_snp.mac_ge2.biallele.LD.structure_analyses.vcf -inputformat VCF -outputfile FINAL_snp.structure -outputformat STRUCTURE

#Edits
#- VCF_PARSER_PLOIDY_QUESTION=HAPLOID
#- VCF_PARSER_POP_QUESTION=true
#- VCF_PARSER_POP_FILE_QUESTION=ind_file.structure
#- VCF_PARSER_PL_QUESTION=false
#- STRUCTURE_WRITER_LOCI_DISTANCE_QUESTION=false
#- STRUCTURE_WRITER_DATA_TYPE_QUESTION=SNP
#- STRUCTURE_WRITER_FAST_FORMAT_QUESTION=false



### Running structure_threader
First generate the params files
```
cd
mkdir Nf_SPANDx_all_seqs_structure_th
module purge
module load anaconda/colsa
export PATH=~/.local/bin:$PATH

cd Nf_SPANDx_all_seqs_structure_th
structure_threader params -o ~/Nf_SPANDx_all_seqs_structure_th/

```
Download the params file to edit and retain a local copy if desired. The default parameters generated by structure_threader in `extraparams` are satisfactory.
#### Edits to `mainparams`
- set
- PLOIDY to 1
- NUMINDS 105 #after filtering out seuqncing reps and high missing data
- NUMLOCI 39849 #SNPs remaining after pgspider
- MISSING -9
- LABEL 1
- POPDATA 1
- MAXPOPS 15 #although this is set by -K on the command line

Run structure_threader (started at 5/27 2p)
```
cd ~/neonectria_genome_reseq_10072020/
sbatch ~/repo/neonectria_genome_reseq_10072020/premise/structure_threader.slurm
```
